otx/algorithms/common/adapters/torch/utils/bs_search_algo.py

```diff
@@ -12,66 +12,199 @@
-    lowest_unavailable_bs = current_bs + 2
-    _, total_mem = torch.cuda.mem_get_info()
+        if max_bs < default_bs:
+            default_bs = max_bs
 
-    def _get_even_center_val(val1, val2):
-        ret = (val1 + val2) // 2
-        if ret % 2 == 1:
-            ret += 1
-        return ret
+        self._train_func = train_func
+        self._default_bs = default_bs
+        self._max_bs = max_bs
+        self._bs_try_history: Dict[int, int] = {}
+        _, self._total_mem = torch.cuda.mem_get_info()
+        self._mem_lower_bound = 0.8 * self._total_mem
+        self._mem_upper_bound = 0.85 * self._total_mem
 
-    while True:
+    def _try_batch_size(self, bs: int) -> Tuple[bool, int]:
         cuda_oom = False
         torch.cuda.reset_max_memory_allocated(device=None)
+        torch.cuda.empty_cache()
 
         try:
-            train_func(current_bs)
+            self._train_func(bs)
         except RuntimeError as e:
             if str(e).startswith("CUDA out of memory."):
                 cuda_oom = True
             else:
                 raise e
 
-        gpu_memory_usage = torch.cuda.max_memory_allocated(device=None) / total_mem
+        max_memory_allocated = torch.cuda.max_memory_allocated(device=None)
+        if not cuda_oom:
+            # Because heapq only supports min heap, use negatized batch size
+            self._bs_try_history[bs] = max_memory_allocated
+
         logger.debug(
-            f"Adapting Batch size => bs : {current_bs}, CUDA_OOM : {cuda_oom}, GPU memory usage : {gpu_memory_usage}%"
+            f"Adapting Batch size => bs : {bs}, CUDA_OOM : {cuda_oom}, "
+            f"GPU memory usage : {max_memory_allocated / self._total_mem}%"
         )
+        torch.cuda.empty_cache()
 
-        # If GPU memory usage is too close to limit, CUDA OOM can be raised during training
-        if cuda_oom or torch.cuda.max_memory_allocated(device=None) >= total_mem * 0.85:
-            if current_bs < lowest_unavailable_bs:
-                lowest_unavailable_bs = current_bs
-            current_bs = _get_even_center_val(current_bs, available_bs)
-        else:
-            available_bs = current_bs
-            current_bs = _get_even_center_val(current_bs, lowest_unavailable_bs)
+        return cuda_oom, max_memory_allocated
 
-        torch.cuda.empty_cache()
-        if lowest_unavailable_bs - available_bs <= 2:
-            break
+    @staticmethod
+    def _get_even_center_val(val1: int, val2: int) -> int:
+        ret = (val1 + val2) // 2
+        if ret % 2 == 1:
+            ret += 1
+        return ret
+
+    def auto_decrease_batch_size(self) -> int:
+        """Decrease batch size if default batch size isn't fit to current GPU device.
+
+        Returns:
+            int: Proper batch size possibly decreased as default value isn't fit
+        """
+        available_bs = 0
+        current_bs = self._default_bs
+        lowest_unavailable_bs = self._default_bs + 2
+
+        while True:
```

Comment:
Maybe we don't need to specify a timeout value for this?

Reply:
I think it always can exit loop  because upper bound decreases as it progresses and also lower bound increases also. And it works with default batch size, so it's doesn't use much time, I think.